corgidrp/detector.py

```diff
@@ -520,7 +528,149 @@ def calc_sat_fwc(emgain_arr,fwcpp_arr,fwcem_arr,sat_thresh):
+        ny = np.arange(0,residual_images[i].shape[0])
+        nxx,nyy = np.meshgrid(nx,ny)
+        nrr = np.sqrt((nxx-rad-5)**2 + (nyy-rad-5)**2)
+        nrr_copy = nrr.copy()
+        nrr_copy[nrr<rad] = residual_images[i][nrr<rad] 
+        nrr_copy[nrr>=rad] = None 
+        ymin = int(cent[i][0])
+        ymax = int(cent[i][1])
+        xmin = int(cent[i][2])
+        xmax = int(cent[i][3])
+
+        bool_innotzero = np.logical_and(nrr<rad,full_residuals[ymin:ymax,xmin:xmax]!=0)
+        bool_iniszero = np.logical_and(nrr<rad,full_residuals[ymin:ymax,xmin:xmax]==0)
+        bool_outisnotzero = np.logical_and(nrr>=rad,full_residuals[ymin:ymax,xmin:xmax]!=0)
+        
+        full_residuals[ymin:ymax,xmin:xmax][bool_innotzero] = (nrr_copy[bool_innotzero] + full_residuals[ymin:ymax,xmin:xmax][bool_innotzero])/2
+        full_residuals[ymin:ymax,xmin:xmax][bool_iniszero] = nrr_copy[bool_iniszero]
+        full_residuals[ymin:ymax,xmin:xmax][bool_outisnotzero] = full_residuals[ymin:ymax,xmin:xmax][bool_outisnotzero]
+
+    if band == '4':
+        full_qe_resel = ndimage.convolve(full_qe, mask4)
+        full_residuals_resel = ndimage.convolve(full_residuals,mask4)
+        percens_resel = (((full_qe_resel - full_residuals_resel)/full_qe_resel)*100)
+    
+    elif band== '1':
+        full_qe_resel = ndimage.convolve(full_qe, mask1)
+        full_residuals_resel = ndimage.convolve(full_residuals,mask1)
+        percens_resel = (((full_qe_resel - full_residuals_resel)/full_qe_resel)*100)
+    
+    full_residuals[full_residuals==0] = None
+    nx = np.arange(0,percens_resel.shape[1])
+    ny = np.arange(0,percens_resel.shape[0])
+    nxx,nyy = np.meshgrid(ny,nx)
+    nrr = np.sqrt((nxx-n/2)**2 + (nyy-n/2)**2)
+    full_residuals[nrr>165]= 1
+    percens_resel[nrr>165] = 0
+    full_qe[nrr>165]= 1
+    full_qe=np.pad(full_qe, ((302,302),(302,302)), mode='constant',constant_values=(1))
+    full_residuals=np.pad(full_residuals, ((302,302),(302,302)), mode='constant',constant_values=(1))
+    percens_resel=np.pad(percens_resel, ((302,302),(302,302)), mode='constant',constant_values=(1))
+    
+    return (full_qe,full_residuals,percens_resel)
+    
+    
+def create_onsky_flatfield(dataset, planet=None,band=None):
+    """Turn this dataset of raster scanned image frames of neptune or uranus into on-sky calibrated flat field 
+    
+        Args:
+            dataset (corgidrp.data.Dataset): a dataset of Image frames (L2a-level)
+            planet (str): neptune or uranus
+            band (str): 1 or 4
+
+    	Returns:
+    		data.FlatField (corgidrp.data.FlatField): a master flat for flat calibration using on sky images of planet in band specified
+    		
+	"""
+    smooth_images=[];raster_images_cent=[];cent=[]; act_cents=[];frames=[]; files=[];
+    planims=planet+'_band_'+band
+    for i in range(len(dataset)):
+        filename=Path(dataset[i].filename).stem.split('_')[0]+'_band_'+Path(dataset[i].filename).stem.split('_')[2]
```

Comment:
I'm not sure we can assume the filename will look exactly like this. What's the purpose of this block? just to exclude files that aren't flat field images that got mistakenly included? If so, I wouldn't worry too much about that for now.

Reply:
The default CGI filenames will look something like CGI_L1_100_0200001001001100001_20270101T120000_000.fits

Reply:
This is intended to pick only those files corresponding to the planet and Filter for which the flatfield has to be created. But I can modify this; I think I can pull the header keywords [Target] and [Filter] and do the same thing.

Reply:
I think you can assume that the dataset passed in is for the specific TARGET and FILTER